readthedocs/doc_builder/builder.py

```diff
@@ -0,0 +1,443 @@
+
+            # TODO: check if `before_build` and `after_build` are still useful
+            # (maybe in commercial?)
+            #
+            # I didn't find they are used anywhere, we should probably remove them
+            before_build.send(
+                sender=self.data.version,
+                environment=self.build_environment,
+            )
+
+            self.pre_system_dependencies()
+            self.system_dependencies()
+            self.post_system_dependencies()
+
+            # Install all ``build.tools`` specified by the user
+            if self.data.config.using_build_tools:
+                self.language_environment.install_build_tools()
+
+            self.pre_create_environment()
+            self.create_environment()
+            self.post_create_environment()
+
+            self.pre_install()
+            self.install()
+            self.post_install()
+
+            # TODO: remove this and document how to do it on `build.jobs.post_install`
+            if self.data.project.has_feature(Feature.LIST_PACKAGES_INSTALLED_ENV):
+                self.language_environment.list_packages_installed()
+
+    def build(self):
+        """
+        Build all the formats specified by the user.
+
+        1. build HTML
+        2. build HTMLZzip
+        3. build PDF
+        4. build ePub
+        """
+        with self.build_environment:
+            self.data.outcomes = defaultdict(lambda: False)
+            self.data.outcomes["html"] = self.build_html()
+            self.data.outcomes["search"] = self.is_type_sphinx()
+            self.data.outcomes["localmedia"] = self.build_htmlzip()
+            self.data.outcomes["pdf"] = self.build_pdf()
+            self.data.outcomes["epub"] = self.build_epub()
+
+            after_build.send(
+                sender=self.data.version,
+            )
+
+    # VCS checkout
+    def pre_checkout(self):
+        # We can't do too much here because we haven't cloned the repository
+        # yet and we don't know what the user wrote in the `.readthedocs.yaml`
+        # yet.
+        #
+        # We could implement something different in the future if we download
+        # the `.readthedocs.yaml` file without cloning.
+        # See https://github.com/readthedocs/readthedocs.org/issues/8935
+        pass
+
+    def checkout(self):
+        log.info(
+            "Clonning repository.",
+        )
+        self.vcs_repository.update()
+
+        identifier = self.data.build_commit or self.data.version.identifier
+        log.info("Checking out.", identifier=identifier)
+        self.vcs_repository.checkout(identifier)
+
+        self.data.config = load_yaml_config(version=self.data.version)
+        self.data.build["config"] = self.data.config.as_dict()
+
+        if self.vcs_repository.supports_submodules:
+            self.vcs_repository.update_submodules(self.data.config)
+
+    def post_checkout(self):
+        commands = []  # self.data.config.build.jobs.post_checkout
```

Comment:
These `config.build.job.*` keys should be the only thing required to implement to make everything work. It may make sense to start this work in another PR branched off from this one so the diff is smaller and easier to review.